src/protocol-handlers/urls.js

```diff
@@ -0,0 +1,147 @@
+const toUri = require('multiaddr-to-uri')
+const { CID } = require('multiformats/cid')
+const fetch = require('node-fetch')
+
+const DEFAULT_GATEWAY = 'https://dweb.link'
+const LOCAL_HOSTNAMES = ['127.0.0.1', '[::1]', '0.0.0.0', '[::]']
+
+/**
+ * @typedef ParsedUrl
+ * @type {object}
+ * @property {string} protocol
+ * @property {string} hostname
+ * @property {string} path
+ */
+
+/**
+ * Parses an IPFS/IPNS/dWeb URL to be handled by IPFS Desktop.
+ *
+ * @param {string} url
+ * @returns {ParsedUrl}
+ */
+function parseUrl (url) {
+  let protocol = null
+  let hostname = null
+  let path = '/'
+
+  if (url.startsWith('ipfs://')) {
+    protocol = 'ipfs'
+    hostname = url.slice(7)
+  } else if (url.startsWith('ipns://')) {
+    protocol = 'ipns'
+    hostname = url.slice(7)
+  } else if (url.startsWith('dweb:/ipfs/')) {
+    protocol = 'ipfs'
+    hostname = url.slice(11)
+  } else if (url.startsWith('dweb:/ipns/')) {
+    protocol = 'ipns'
+    hostname = url.slice(11)
+  } else {
+    return null
+  }
+
+  if (hostname.includes('/')) {
+    const [first, ...rest] = hostname.split('/')
+    hostname = first
+    path = '/' + rest.join('/')
+  }
+
+  return { protocol, hostname, path }
+}
+
+async function getPublicGatewayUrl (ctx) {
+  if (!ctx.webui) {
+    // Best effort. If the Web UI window wasn't created yet, we just return the default
+    // gateway.
+    return DEFAULT_GATEWAY
+  }
+
+  return await ctx.webui
+    .webContents
+    .executeJavaScript('JSON.parse(localStorage.getItem("ipfsPublicGateway")) || "https://dweb.link"', true)
```

Comment:
instead of the OR in the executed javascript, can we do the OR with something like

```js

/** @type {string} */
let publicGatewayUrl
try {
   publicGatewayUrl = await ctx.webui.webContents.executeJavaScript('JSON.parse(localStorage.getItem("ipfsPublicGateway"))', true))
} catch {
  publicGatewayUrl = 'https://dweb.link'
}

return publicGatewayUrl
```

Reply:
Try catch won't work here. Do you think this would be an improvement:

```javascript
/** @type {string|null} */
  let publicGatewayUrl = await ctx.webui.webContents.executeJavaScript('JSON.parse(localStorage.getItem("ipfsPublicGateway"))')
  if (!publicGatewayUrl) {
    publicGatewayUrl = 'https://dweb.link'
  }
```
?

Reply:
I think that works, along with wrapping *any and all* JSON.parse, and awaits, in try/catch blocks.